### PR TITLE
Add IsSystemView and IsIgnoredView to enable future H2 version 2 changes

### PR DESCRIPTION
**Pull request by jonknight73**

### CODE REVIEW

File: morf-h2/src/main/java/org/alfasoftware/morf/jdbc/h2/H2Dialect.java [hunk 1]
```diff
@@ -597,13 +596,10 @@ protected String getSqlForRowNumber() {
...
-      case ROW_NUMBER:
-        return "ROW_NUMBER()";
-
-      default:
-        return super.getSqlForWindowFunction(function);
+    if (functionType == FunctionType.ROW_NUMBER) {
+      return "ROW_NUMBER()";
     }
+    return super.getSqlForWindowFunction(function);
   }
```
Comment (jsimlo) [inline morf-h2/src/main/java/org/alfasoftware/morf/jdbc/h2/H2Dialect.java:603]: We want a switch statement here.
That would be consistent with other similar functions, where more cases can be expected in the future.

**Review by jsimlo — CHANGES_REQUESTED** [review]

File: morf-h2/src/main/java/org/alfasoftware/morf/jdbc/h2/H2Dialect.java (hunk 1 above)
Comment (jonknight73) [reply morf-h2/src/main/java/org/alfasoftware/morf/jdbc/h2/H2Dialect.java:603]: I was thinking that initially, but should it be an if statement until there is actually enough options to merit a switch statement?

File: morf-core/src/test/java/org/alfasoftware/morf/xml/DummyXmlOutputStreamProvider.java [hunk 2]
```diff
@@ -77,7 +77,7 @@ public void clearDestination() {
    * @return Convert the output to a String
    */
   public String getXmlString() {
-    return new String(testOutputStream.toByteArray());
+    return testOutputStream.toString();
```
Comment (jsimlo) [inline morf-core/src/test/java/org/alfasoftware/morf/xml/DummyXmlOutputStreamProvider.java:80]: Are we sure about this?
To bytes and back was potentially doing some character encoding conversions.
Perhaps this is better? Perhaps this is a functional change?

File: morf-integration-test/src/test/resources/log4j.properties [hunk 3]
```diff
@@ -9,3 +9,5 @@ log4j.appender.CONSOLE.layout.ConversionPattern=%d{yyyy-MM-dd HH:mm:ss} %-5p %c{
 log4j.logger.org.alfasoftware.morf.diagnostics.DatabaseAndSystemTimeConsistencyChecker=DEBUG
 
 log4j.logger.org.alfasoftware.morf.jdbc.AbstractConnectionResources=INFO
+
+log4j.logger.org.alfasoftware.morf.jdbc.DatabaseMetaDataProvider=DEBUG
```
Comment (jsimlo) [inline morf-integration-test/src/test/resources/log4j.properties:13]: Was this an intentional change?

File: morf-postgresql/src/main/java/org/alfasoftware/morf/jdbc/postgresql/PostgreSQLDialect.java [hunk 4]
```diff
@@ -694,7 +695,7 @@ public Collection<String> alterTableChangeColumnStatements(Table table, Column o
 
     boolean alterNullable = oldColumn.isNullable() != newColumn.isNullable();
     boolean alterType = oldColumn.getType() != newColumn.getType() || oldColumn.getScale() != newColumn.getScale() || oldColumn.getWidth() != newColumn.getWidth();
-    boolean alterDefaultValue = oldColumn.getDefaultValue() != newColumn.getDefaultValue();
+    boolean alterDefaultValue = !Objects.equals(oldColumn.getDefaultValue(), newColumn.getDefaultValue());
```
Comment (jsimlo) [inline morf-postgresql/src/main/java/org/alfasoftware/morf/jdbc/postgresql/PostgreSQLDialect.java:698]: Note: `java.util.Objects.equals()` goes further than `!=`, and uses `Object.equals()`.
Therefore this is a potential behaviour change and must be investigated.

`Column.getDefaultValue()` returns a String. Objects.equals sounds like a better choice.

File: morf-core/src/test/java/org/alfasoftware/morf/metadata/TestSchemaUtils.java [hunk 5]
```diff
@@ -85,8 +86,8 @@ public void testCopySchemaWithExclusions() {
     //...assert during the copy excluded tables and views are being removed
     assertEquals("2 tables should exist", 2, schema.tables().size());
     assertEquals("1 view should exist", 1, schema.views().size());
-    assertFalse(schema.tables().stream().anyMatch(t -> t.getName() == excludePrefix));
-    assertFalse(schema.views().stream().anyMatch(t -> t.getName() == excludeMatch));
+    assertFalse(schema.tables().stream().anyMatch(t -> Objects.equals(t.getName(), excludePrefix)));
+    assertFalse(schema.views().stream().anyMatch(t -> Objects.equals(t.getName(), excludeMatch)));
```
Comment (jsimlo) [inline morf-core/src/test/java/org/alfasoftware/morf/metadata/TestSchemaUtils.java:90]: Note: `java.util.Objects.equals()` goes further than `==`, and uses `Object.equals()`.
Therefore this is a potential behaviour change and must be investigated.

We are comparing Strings. Objects.equals sounds like a better choice.

File: morf-excel/src/test/java/org/alfasoftware/morf/excel/TestTableOutputter.java [hunk 6]
```diff
@@ -79,7 +79,7 @@ public void testTableOutputWithFieldsPopulated() throws WriteException {
     // Given
     Record record1 = mock(Record.class);
     when(record1.getString("Col1")).thenReturn("STRING VALUE");
-    when(record1.getBigDecimal("Col2")).thenReturn(new BigDecimal(12.34));
+    when(record1.getBigDecimal("Col2")).thenReturn(BigDecimal.valueOf(12.34));
```
Comment (jsimlo) [inline morf-excel/src/test/java/org/alfasoftware/morf/excel/TestTableOutputter.java:82]: Note: `new BigDecimal(double)` differs from `BigDecimal.valueOf())`, this is a potential change!

This is a test, so we are fine.

File: morf-postgresql/src/main/java/org/alfasoftware/morf/jdbc/postgresql/PostgreSQLDialect.java [hunk 7]
```diff
@@ -713,13 +714,10 @@ public Collection<String> alterTableChangeColumnStatements(Table table, Column o
...
-                + (alterType ? " ALTER COLUMN " + newColumn.getName() + " TYPE " + sqlRepresentationOfColumnType(newColumn, false, false, true) : "")
-                + (alterDefaultValue && (alterNullable || alterType) ? "," : "")
-                + (alterDefaultValue ? " ALTER COLUMN " + newColumn.getName() + (!newColumn.getDefaultValue().isEmpty() ? " SET DEFAULT " + sqlForDefaultClauseLiteral(newColumn) : " DROP DEFAULT") : "")
-        );
+    sqlBuilder.append("ALTER TABLE ").append(schemaNamePrefix(table)).append(table.getName())
+            .append(alterNullable ? " ALTER COLUMN " + newColumn.getName() + (newColumn.isNullable() ? " DROP NOT NULL" : " SET NOT NULL") : "")
+            .append(alterNullable && alterType ? "," : "").append(alterType ? " ALTER COLUMN " + newColumn.getName() + " TYPE " + sqlRepresentationOfColumnType(newColumn, false, false, true) : "")
+            .append(alterDefaultValue && (alterNullable || alterType) ? "," : "").append(alterDefaultValue ? " ALTER COLUMN " + newColumn.getName() + (!newColumn.getDefaultValue().isEmpty() ? " SET DEFAULT " + sqlForDefaultClauseLiteral(newColumn) : " DROP DEFAULT") : "");
     return sqlBuilder.toString();
```
Comment (jsimlo) [inline morf-postgresql/src/main/java/org/alfasoftware/morf/jdbc/postgresql/PostgreSQLDialect.java:721]: I think we should revert this change, and drop the `StringBuilder` completely instead.
The mix of appends with conditional concats inside is probably worse than before.

File: morf-postgresql/src/main/java/org/alfasoftware/morf/jdbc/postgresql/PostgreSQLDialect.java [hunk 8]
```diff
@@ -796,7 +794,6 @@ public void prepareStatementParameters(NamedParameterPreparedStatement statement
 
       default:
         super.prepareStatementParameters(statement, values, parameter);
-        return;
```
Comment (jsimlo) [inline morf-postgresql/src/main/java/org/alfasoftware/morf/jdbc/postgresql/PostgreSQLDialect.java:799]: Cannot drop this due to our general PMD rules.
Need a `return` (preferred) or `break` (but then all cases should do `break` for consistency).

File: morf-postgresql/src/main/java/org/alfasoftware/morf/jdbc/postgresql/PostgreSQLDialect.java [hunk 9]
```diff
@@ -824,27 +821,26 @@ protected String getSqlForEvery(Function function) {
...
-    }
-    sqlBuilder.append(getSqlFrom(selectStatement.build()));
+      SelectStatementBuilder selectStatement = select().fields(field("ctid")).from(statement.getTable());
+      if (statement.getWhereCriterion() != null) {
+        selectStatement = selectStatement.where(statement.getWhereCriterion());
+      }
+      sqlBuilder.append(getSqlFrom(selectStatement.build()));
 
-    sqlBuilder.append(" LIMIT " + statement.getLimit().get() + ")");
+      sqlBuilder.append(" LIMIT ").append(statement.getLimit().get()).append(")"); //NOSONAR
```
Comment (jsimlo) [inline morf-postgresql/src/main/java/org/alfasoftware/morf/jdbc/postgresql/PostgreSQLDialect.java:839]: What was the NOSONAR for?

File: morf-testsupport/src/main/java/org/alfasoftware/morf/xml/DummyXmlOutputStreamProvider.java [hunk 10]
```diff
@@ -77,7 +77,7 @@ public void clearDestination() {
    * @return Convert the output to a String
    */
   public String getXmlString() {
-    return new String(testOutputStream.toByteArray());
+    return testOutputStream.toString();
```
Comment (jsimlo) [inline morf-testsupport/src/main/java/org/alfasoftware/morf/xml/DummyXmlOutputStreamProvider.java:80]: Are we sure about this?
To bytes and back was potentially doing some character encoding conversions.
Perhaps this is better? Perhaps this is a functional change?

**Review by jsimlo — CHANGES_REQUESTED** [review]

File: morf-core/src/test/java/org/alfasoftware/morf/xml/DummyXmlOutputStreamProvider.java (hunk 2 above)
Comment (jonknight73) [reply morf-core/src/test/java/org/alfasoftware/morf/xml/DummyXmlOutputStreamProvider.java:80]: Reverted, it needs a proper look to verify the encoding treatment.

File: morf-core/src/test/java/org/alfasoftware/morf/metadata/TestSchemaUtils.java (hunk 5 above)
Comment (jonknight73) [reply morf-core/src/test/java/org/alfasoftware/morf/metadata/TestSchemaUtils.java:90]: This is a test where comparing the value rather than the pointer is clearly the intended behaviour.

File: morf-postgresql/src/main/java/org/alfasoftware/morf/jdbc/postgresql/PostgreSQLDialect.java (hunk 7 above)
Comment (jonknight73) [reply morf-postgresql/src/main/java/org/alfasoftware/morf/jdbc/postgresql/PostgreSQLDialect.java:721]: Done.

File: morf-postgresql/src/main/java/org/alfasoftware/morf/jdbc/postgresql/PostgreSQLDialect.java (hunk 8 above)
Comment (jonknight73) [reply morf-postgresql/src/main/java/org/alfasoftware/morf/jdbc/postgresql/PostgreSQLDialect.java:799]: Restored.

File: morf-postgresql/src/main/java/org/alfasoftware/morf/jdbc/postgresql/PostgreSQLDialect.java (hunk 9 above)
Comment (jonknight73) [reply morf-postgresql/src/main/java/org/alfasoftware/morf/jdbc/postgresql/PostgreSQLDialect.java:839]: Done

File: morf-testsupport/src/main/java/org/alfasoftware/morf/xml/DummyXmlOutputStreamProvider.java (hunk 10 above)
Comment (jonknight73) [reply morf-testsupport/src/main/java/org/alfasoftware/morf/xml/DummyXmlOutputStreamProvider.java:80]: Reverted

File: morf-integration-test/src/test/resources/log4j.properties (hunk 3 above)
Comment (jonknight73) [reply morf-integration-test/src/test/resources/log4j.properties:13]: It was an attempt to increase code coverage. It failed. And I meant to revert it.

**Review by jsimlo — APPROVED** [review]

**Comment by jonknight73** [comment]
I would merge PRs 270 and 271 first. This one is huge, and will likely conflict the others.
